util/tracker.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+)
+
+// MasterTracker tracks the current mastership state across multiple IDs.
+type MasterTracker struct {
+	mu          sync.RWMutex
+	masterFor   map[int64]bool
+	masterCount int
+}
+
+// NewMasterTracker creates a new MasterTracker instance to track the mastership
+// status for the given set of ids.
+func NewMasterTracker(ids []int64) *MasterTracker {
+	mf := make(map[int64]bool)
+	for _, id := range ids {
+		mf[id] = false
+	}
+	return &MasterTracker{masterFor: mf}
+}
+
+// Set changes the tracked mastership status for the given id.  This method should
+// be called exactly once for each state transition.
+func (mt *MasterTracker) Set(id int64, val bool) {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	existing, ok := mt.masterFor[id]
+	if ok && val != !existing {
+		glog.Errorf("toggle masterFor[%d] from %v to %v!", id, existing, val)
```

Comment:
I think returning an error would make life much harder for the caller -- it's not really going to be able to do anything other than log it.  I have downgraded to a `Warningf` -- the function can still do what it's asked in this case, it's just a breach of an expected precondition.